File: Solutions/Dungeon_Game/Dungeon_Game.py
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        r = len(dungeon)
        c = len(dungeon[0])
        ans = [[0]*c for i in range(r)]
        ans[r-1][c-1] = 1 if dungeon[r-1][c-1] > 0 else (1 - dungeon[r-1][c-1])
        
        for i in range(r-2, -1, -1):
            ans[i][c-1] = max(ans[i+1][c-1] - dungeon[i][c-1], 1)
        for j in range(c-2, -1, -1):
            ans[r-1][j] = max(ans[r-1][j+1] - dungeon[r-1][j], 1)
            
        for i in range(r-2, -1, -1):
            for j in range(c-2, -1, -1):
                ans[i][j] = max(min(ans[i+1][j], ans[i][j+1]) - dungeon[i][j], 1)
                
        return ans[0][0]
        
#Fastest:
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        if not dungeon:
            return 0
        
        num_row, num_col = len(dungeon), len(dungeon[0])
        dp = [float('inf')] * num_col
        dp[num_col - 1] = 1
        
        for row in reversed(range(num_row)):
            for col in reversed(range(num_col)):
                if row == num_row - 1 and col == num_col - 1:
                    dp[col] = max(1, 1 - dungeon[row][col])
                elif row == num_row - 1:
                    dp[col] = max(1, dp[col + 1] - dungeon[row][col])
                elif col == num_col - 1:
                    dp[col] = max(1, dp[col] - dungeon[row][col])
                else:
                    dp[col] = max(1, min(dp[col], dp[col + 1]) - dungeon[row][col])
        
        return dp[0]
        

File: Solutions/Dungeon_Game/Dungeon_Game.cpp
class Solution {
public:
    int calculateMinimumHP(vector<vector<int>>& dungeon) {
        int rows = dungeon.size();
        int cols = dungeon[0].size();
        
        vector<vector<int>> ans(rows, vector<int>(cols));
        
        ans[rows-1][cols-1] = dungeon[rows-1][cols-1] > 0 ? 1 : 1 - dungeon[rows-1][cols-1];
            
        for(int i = rows-2; i >= 0; i--)
            ans[i][cols-1] = max(ans[i+1][cols-1] - dungeon[i][cols-1], 1);
        for(int j = cols-2; j >= 0; j--)
            ans[rows-1][j] = max(ans[rows-1][j+1] - dungeon[rows-1][j], 1);
        
        for(int i = rows-2; i >= 0; i--)
            for(int j = cols-2; j >=0; j--)
                ans[i][j] = max(min(ans[i+1][j], ans[i][j+1]) - dungeon[i][j], 1);
        
        return ans[0][0];
    }
};

/*
Fastest:
static auto fast=[]{ios_base::sync_with_stdio(false);cin.tie(nullptr);return 0;}();
class Solution {
public:
    int calculateMinimumHP(vector<vector<int>>& D) {
        int m = D.size();
        if(m==0)
        {
            return 0;
        }
        int n = D[0].size();
        if(n==0)
        {
            return 0;
        }
        for(int i=m-1;i>=0;i--)
        {
            for(int j=n-1;j>=0;j--)
            {
                if(i==m-1 && j==n-1)
                {
                    if(D[i][j]>=0)
                    {
                        D[i][j] = 1;
                    }
                    else{
                        D[i][j] = (-1*D[i][j]) + 1;
                    }
                    continue;
                }
                if(i==m-1)
                {
                    if(D[i][j]>=0){
                    D[i][j]= max(1,D[i][j+1]-D[i][j]);
                    }else{
                      D[i][j]= (-1*D[i][j]) + D[i][j+1];
                    }
                    continue;
                }
                 if(j==n-1)
                {
                   if(D[i][j]>=0){
                    D[i][j]= max(1,D[i+1][j]-D[i][j]);
                    }else{
                       D[i][j]= (-1*D[i][j]) + D[i+1][j];
                    }
                    continue;
                }
                if(D[i][j]>=0)
                {
                    D[i][j]= min(max(1,D[i][j+1]-D[i][j]),max(1,D[i+1][j]-D[i][j]));
                }
                else{
                 D[i][j]= (-1*D[i][j]) + min(D[i+1][j],D[i][j+1]);   
                }
                
                
            }
        }
        return D[0][0];
    }
};
*/